client/checkout/classic/upe-deferred-intent-creation/event-handlers.js

```diff
@@ -109,8 +109,55 @@ jQuery( function ( $ ) {
 		) {
 			for ( const upeElement of $( '.wcpay-upe-element' ).toArray() ) {
 				await mountStripePaymentElement( api, upeElement );
+				restrictPaymentMethodToLocation( upeElement );
 			}
 			maybeEnableStripeLink( api );
 		}
 	}
+
+	/**
+	 * Hides payment method if it has set specific countries in the PHP class.
+	 *
+	 * @param {Object} upeElement The selector of the DOM element of particular payment method to mount the UPE element to.
+	 **/
+	function restrictPaymentMethodToLocation( upeElement ) {
+		const paymentMethodsConfig = getUPEConfig( 'paymentMethodsConfig' );
+		const paymentMethodType = upeElement.dataset.paymentMethodType;
+		const isRestrictedInAnyCountry = !! paymentMethodsConfig[
+			paymentMethodType
+		].countries.length;
+
+		if ( isRestrictedInAnyCountry ) {
+			togglePaymentMethodForCountry(
+				paymentMethodType,
+				paymentMethodsConfig[ paymentMethodType ].countries
+			);
+			$( '#billing_country' ).on( 'change', function () {
```

Comment:
If I understand correctly there's a chance that we add this event listener more than once. I'd suggest to attach the event listener only once, maybe using a flag that can be flipped when the first event listener is attached.

Reply:
Yes, there is that chance. When we initially implemented this in Split UPE files, I specifically asked whether we should do this optimization or stick with a more straightforward solution, and the consensus was to attach the listener multiple times since the performance gain is so tiny compared to the effect on the codebase.

Reply:
Got it thanks, marking this as resolved